ffi/examples/visit-expression/engine_to_kernel_expression.h

```diff
@@ -0,0 +1,315 @@
+#pragma once
+
+#include "delta_kernel_ffi.h"
+#include "expression.h"
+#include <assert.h>
+#include <stdlib.h>
+
+/**
+ * This module converts an engine expression (ExpressionItemList) back into
+ * a kernel expression (SharedExpression) by calling the appropriate visit_*
+ * functions from KernelExpressionVisitorState.
+ */
+
+// Forward declarations
+uintptr_t engine_to_kernel_expression_item(
+    KernelExpressionVisitorState* state,
+    ExpressionItem item);
+
+uintptr_t engine_to_kernel_literal(
+    KernelExpressionVisitorState* state,
+    struct Literal* lit) {
+  switch (lit->type) {
+    case Integer:
+      return visit_expression_literal_int(state, lit->value.integer_data);
+    case Long:
+      return visit_expression_literal_long(state, lit->value.long_data);
+    case Short:
+      return visit_expression_literal_short(state, lit->value.short_data);
+    case Byte:
+      return visit_expression_literal_byte(state, lit->value.byte_data);
+    case Float:
+      return visit_expression_literal_float(state, lit->value.float_data);
+    case Double:
+      return visit_expression_literal_double(state, lit->value.double_data);
+    case String: {
+      KernelStringSlice str_slice = {
+        .ptr = lit->value.string_data,
+        .len = strlen(lit->value.string_data)
+      };
+      ExternResultusize result = visit_expression_literal_string(
+          state, str_slice, NULL);
+      return result.ok;
+    }
+    case Boolean:
+      return visit_expression_literal_bool(state, lit->value.boolean_data);
+    case Timestamp:
+      return visit_expression_literal_timestamp(state, lit->value.long_data);
+    case TimestampNtz:
+      return visit_expression_literal_timestamp_ntz(state,
+          lit->value.long_data);
+    case Date:
+      return visit_expression_literal_date(state, lit->value.integer_data);
+    case Binary: {
+      return visit_expression_literal_binary(
+          state, lit->value.binary.buf, lit->value.binary.len);
+    }
+    case Decimal: {
+      struct Decimal* dec = &lit->value.decimal;
+      return visit_expression_literal_decimal(
+          state, dec->hi, dec->lo, dec->precision, dec->scale);
+    }
+    case Null:
+      return visit_expression_literal_null(state);
+    case Struct:
+      fprintf(stderr, "Error: Struct literal type not supported\n");
+      assert(0 && "Struct literal type not supported");
+      abort(); // Explicitly abort even if assertions are disabled
```

Comment:
Maybe unnecessary to manually abort since we only ever run these with assertions enabled.

Reply:
I think generally we attempt not to bring down the whole system (`panic!` in rust) and instead bias towards returning Error types. perhaps this is an exception but wondering if we could do better (in follow-up?)